diff --git a/portable/GCC/IA32_flat/ISR_Support.h b/portable/GCC/IA32_flat/ISR_Support.h
index 18e10bf91..cbc9dd0fd 100644
--- a/portable/GCC/IA32_flat/ISR_Support.h
+++ b/portable/GCC/IA32_flat/ISR_Support.h
@@ -113,7 +113,7 @@
     .endif
 
     2:
-    popa
+    //popa
 
 .endm
 /*-----------------------------------------------------------*/
diff --git a/portable/GCC/IA32_flat/portASM.S b/portable/GCC/IA32_flat/portASM.S
index c01b32bce..88f3b50ef 100644
--- a/portable/GCC/IA32_flat/portASM.S
+++ b/portable/GCC/IA32_flat/portASM.S
@@ -213,6 +213,8 @@ _skip_context_switch:
         1:
     .endif
 
+    movb $0x20, %al
+    outb %al, $0x20
     popa
 
     /* EOI. */
diff --git a/portable/GCC/IA32_flat/portmacro.h b/portable/GCC/IA32_flat/portmacro.h
index 085250e04..a6c308fd2 100644
--- a/portable/GCC/IA32_flat/portmacro.h
+++ b/portable/GCC/IA32_flat/portmacro.h
@@ -288,10 +288,14 @@ BaseType_t xPortInstallInterruptHandler( ISR_Handler_t pxHandler,
 #define portAPIC_MAX_SUB_PRIORITY      ( 0x0fUL )
 #define portMAX_API_CALL_PRIORITY      ( ( configMAX_API_CALL_INTERRUPT_PRIORITY << portAPIC_PRIORITY_SHIFT ) | portAPIC_MAX_SUB_PRIORITY )
 
-/* Asserts if interrupt safe FreeRTOS functions are called from a priority
- * above the max system call interrupt priority. */
-#define portAPIC_PROCESSOR_PRIORITY    ( *( ( volatile uint32_t * ) ( configAPIC_BASE + 0xA0UL ) ) )
-#define portASSERT_IF_INTERRUPT_PRIORITY_INVALID()    configASSERT( ( portAPIC_PROCESSOR_PRIORITY ) <= ( portMAX_API_CALL_PRIORITY ) )
+#ifndef portASSERT_IF_INTERRUPT_PRIORITY_INVALID
+    #if ( configMAX_API_CALL_INTERRUPT_PRIORITY == portMAX_PRIORITY )
+        #define portASSERT_IF_INTERRUPT_PRIORITY_INVALID()
+    #else
+        #define portAPIC_PROCESSOR_PRIORITY                   ( *( ( volatile uint32_t * ) ( configAPIC_BASE + 0xA0UL ) ) )
+        #define portASSERT_IF_INTERRUPT_PRIORITY_INVALID()    configASSERT( ( portAPIC_PROCESSOR_PRIORITY ) <= ( portMAX_API_CALL_PRIORITY ) )
+    #endif
+#endif
 
 /* *INDENT-OFF* */
 #ifdef __cplusplus
